contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,460 @@
+      * `PVName` is `PV.name` of the attached PersistentVolume.
+      * `NodeName` is `Node.name` of the node where the volume should be attached to.
+    * If a `VolumeAttachment` object with the corresponding name already exists, the in-tree volume plugin will simply begin to poll it as defined below. The object is not modified; only the external-attacher should change the status fields; and the external-attacher is responsible for it’s own retry and error handling logic.
+2. Poll the `VolumeAttachment` object waiting for one of the following conditions:
+    * The `VolumeAttachment.Status.Attached` field to become `true`.
+      * The operation completes successfully.
+    * An error to be set in the `VolumeAttachment.Status.AttachError` field.
+      * The operation terminates with the specified error.
+    * The operation to timeout.
+      * The operation terminates with timeout error.
+    * The `VolumeAttachment.DeletionTimestamp` is set.
+      * The operation terminates with an error indicating a detach operation is in progress.
+      * The `VolumeAttachment.Status.Attached` value must not be trusted. The attach/detach controller has to wait until the object is deleted by the external-attacher before creating a new instance of the object.
+
+When the controller decides to detach a CSI volume, it will call the in-tree CSI volume plugin’s detach method. The in-tree CSI volume plugin’s detach method will do the following:
+
+1. Delete the corresponding `VolumeAttachment` object (defined in the “Communication Channels” section) to indicate the volume should be detached.
+2. Poll the `VolumeAttachment` object waiting for one of the following conditions:
+    * The `VolumeAttachment.Status.Attached` field to become false.
+      * The operation completes successfully.
+    * An error to be set in the `VolumeAttachment.Status.DetachError` field.
+      * The operation terminates with the specified error.
+    * The object to no longer exists.
+      * The operation completes successfully.
+    * The operation to timeout.
+      * The operation terminates with timeout error.
+
+### Recommended Mechanism for Deploying CSI Drivers on Kubernetes
+
+Although, Kubernetes does not dictate the packaging for a CSI volume driver, it offers the following recommendations to simplify deployment of a containerized CSI volume driver on Kubernetes.
+
+![Recommended CSI Deployment Diagram](container-storage-interface_diagram1.png?raw=true "Recommended CSI Deployment Diagram")
+
+To deploy a containerized third-party CSI volume driver, it is recommended that storage vendors:
+
+  * Create a “CSI volume driver” container that implements the volume plugin behavior and exposes a gRPC interface via a unix domain socket, as defined in the CSI spec (including Controller, Node, and Identity services).
+  * To deploy a CSI plugin, a cluster admin should deploy the following
```

Comment:
This ends, but doesn't have a sub-list?

Reply:
I swapped two lines in the bullet list so it makes some sense. @saad-ali please check. The structure is not clear to me.

> ... it is recommended that storage vendors:
>  * The Kubernetes team will provide ..

that does not say anything about storage vendors

Reply:
Clarifying this.